baseplate/frameworks/queue_consumer/kombu.py

```diff
@@ -127,6 +176,61 @@ def handle(self, message: kombu.Message) -> None:
             message.ack()
 
 
+class KombuBatchMessageHandler(MessageHandler):
+    def __init__(
+        self,
+        baseplate: Baseplate,
+        name: str,
+        handler_fn: BatchHandler,
+        error_handler_fn: Optional[BatchErrorHandler] = None,
+    ):
+        self.baseplate = baseplate
+        self.name = name
+        self.handler_fn = handler_fn
+        self.error_handler_fn = error_handler_fn
+
+    def handle(self, messages: Sequence[kombu.Message]) -> None:  # pylint: disable=arguments-differ
+        logger.info("Processing batch with %i messages", len(messages))
+
+        context = self.baseplate.make_context_object()
+        try:
+            # We place the call to ``baseplate.make_server_span`` inside the
+            # try/except block because we still want Baseplate to see and
+            # handle the error (publish it to error reporting)
+            with self.baseplate.make_server_span(context, self.name) as span:
+                span.set_tag("kind", "batch_consumer")
+
+                message: kombu.Message
+                for message in messages:
+                    delivery_info = message.delivery_info
+                    span.set_tag("amqp.routing_key", delivery_info.get("routing_key", ""))
+                    span.set_tag("amqp.consumer_tag", delivery_info.get("consumer_tag", ""))
+                    span.set_tag("amqp.delivery_tag", delivery_info.get("delivery_tag", ""))
+                    span.set_tag("amqp.exchange", delivery_info.get("exchange", ""))
+
+                self.handler_fn(context, list(messages))
+        except Exception as exc:  # pylint: disable=broad-except
+            logger.exception(
+                "Unhandled error while trying to process a message.  The message "
+                "has been returned to the queue broker."
+            )
+            if self.error_handler_fn:
+                self.error_handler_fn(context, messages, exc)
+            else:
+                for message in messages:
+                    if not message.acknowledged:
+                        message.requeue()
+
+            if isinstance(exc, FatalMessageHandlerError):
+                logger.info("Recieved a fatal error, terminating the server.")
+                raise
+        else:
+            for message in messages:
+                if not message.acknowledged:
+                    message.ack()
```

Comment:
🔕  I wonder if we should actually do this for batch messages? It might make more sense to have the handler_fn call `ack` itself rather than doing it here?  My thinking is that it makes handling errors tricky if developers are expecting the batch to be `ack`-ed at this level so they aren't doing it in their handler and they raise an exception 1/2 way through processing a batch, we would re-queue all the messages since none were `ack`-ed but we would be re-queueing a bunch that had already been processed?

Reply:
I agree with the conceptual questions. The reason why I implemented this behaviour is that the single `KombuMessageHandler` also has the same type of `else` clause where the message it has processed is acknowledged. However, that one is actually not even checking whether the message had been acknowledged at all, so maybe we should add that check?

I think that clients should be aware that the default behaviour is that unack'ed messages will be ack'ed if processing completes successfully and that all unacknowledged messages are requeued upon failure. However, clients should also be allowed to ack, requeue and discard messages themselves, for more fine grained control.